libsolidity/formal/BMC.cpp

```diff
@@ -290,104 +291,214 @@ bool BMC::visit(Conditional const& _op)
+		{
+			if (loopControl.kind == LoopControlKind::Break)
+			{
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
+				broke = broke || loopControl.pathConditions;
+			}
+			else if (loopControl.kind == LoopControlKind::Continue)
+			{
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
+				continues = continues || loopControl.pathConditions;
+			}
+		}
+
+		// not checking loop condition to merge variables
+		// if there are no paths to breaks or continue statements, do-while loop body is fully executed at least once
+		mergeVariables(!broke && !continues, indicesAfter, copyVariableIndices());
+
+		loopScopes.pop();
+		start = 1;
 	}
-	else
+
+	_node.condition().accept(*this);
+
+	unsigned int bmcLoopIterations = m_settings.bmcLoopIterations.value_or(1);
+	for (unsigned int i = start; i < bmcLoopIterations; ++i)
 	{
-		_node.condition().accept(*this);
-		if (isRootFunction())
-			addVerificationTarget(
-				VerificationTargetType::ConstantCondition,
-				expr(_node.condition()),
-				&_node.condition()
-			);
+		loopScopes.emplace();
 
-		indicesAfterLoop = visitBranch(&_node.body(), expr(_node.condition())).first;
-	}
+		auto indicesAfter = visitBranch(&_node.body(), expr(_node.condition())).first;
 
-	// We reset the execution to before the loop
-	// and visit the condition in case it's not a do-while.
-	// A do-while's body might have non-precise information
-	// in its first run about variables that are touched.
-	resetVariableIndices(indicesBeforeLoop);
-	if (!_node.isDoWhile())
-		_node.condition().accept(*this);
+		smtutil::Expression continues(false);
+		for (auto const& loopControl: loopScopes.top())
+		{
+			if (loopControl.kind == LoopControlKind::Break)
+			{
+				// use SSAs associated with this break statement only if
+				// loop didn't break or continue before
+				// loop condition is included in 'break' path conditions
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
+				broke = broke || loopControl.pathConditions;
+			}
+			else if (loopControl.kind == LoopControlKind::Continue)
+			{
+				// use SSAs associated with this continue statement only if
+				// loop didn't break or continue before
+				// loop condition is included in continue path conditions
+				mergeVariables(
+					!broke && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
+				);
+				continues = continues || loopControl.pathConditions;
+			}
+		}
 
-	mergeVariables(expr(_node.condition()), indicesAfterLoop, copyVariableIndices());
+		mergeVariables(
+			!broke && !continues && expr(_node.condition()),
+			indicesAfter,
+			copyVariableIndices()
+		);
+		loopScopes.pop();
+		_node.condition().accept(*this);
```

Comment:
This could be moved to the beginning of the loop, and removed from before the loop.

Reply:
Right, also the code seems to be wrong. Currently do-while always accepts the condition at least once, and it shouldn't happen when do-while is broken in the first iteration.